core/common/assets/js/api/utils/data.js

```diff
@@ -0,0 +1,234 @@
+import Cache from './data/cache';
+
+// TODO: Return it from the server. Original at WP_REST_Server.
+export const READABLE = [ 'GET' ],
+	CREATABLE = [ 'POST' ],
+	EDITABLE = [ 'POST', 'PUT', 'PATCH' ],
+	DELETABLE = [ 'DELETE' ],
+	ALLMETHODS = [ 'GET', 'POST', 'PUT', 'PATCH', 'DELETE' ];
+
+export default class Data {
+	constructor( args = {} ) {
+		this.args = Object.assign( args, {
+			namespace: 'elementor',
+			version: '1',
+		} );
+
+		this.cacheStroage = new Cache( this );
+
+		this.baseEndpointAddress = '';
+
+		elementorCommon.elements.$window.on( 'elementor:loaded', this.onElementorLoaded.bind( this ) );
+	}
+
+	onElementorLoaded() {
+		const { namespace, version } = this.args;
+
+		this.baseEndpointAddress = `${ elementor.config.home_url }/wp-json/${ namespace }/v${ version }/`;
+	}
+
+	getHTTPMethod( type ) {
+		switch ( type ) {
+			case 'create':
+				return 'POST';
+
+			case 'delete':
+				return 'DELETE';
+
+			case 'get':
+				return 'GET';
+
+			case 'update':
+				return 'PUT';
+		}
+
+		return false;
+	}
+
+	getAllowedMethods( type ) {
+		switch ( type ) {
+			case 'create':
+				return CREATABLE;
+
+			case 'delete':
+				return DELETABLE;
+
+			case 'get':
+				return READABLE;
+
+			case 'update':
+				return EDITABLE;
+		}
+
+		return false;
+	}
+
+	/**
+	 * @param {string} type
+	 * @param {string} command
+	 * @param {{}} args
+	 *
+	 * @returns {string} Endpoint
+	 */
+	commandToEndpoint( type, command, args ) {
+		let endPoint = command;
+
+		if ( endPoint.includes( 'index' ) ) {
+			endPoint = endPoint.replace( '/index', '' );
+		}
+
+		if ( args.query.id ) {
+			endPoint += '/' + args.query.id.toString();
+
+			delete args.query.id;
+		}
+
+		delete args.query.type;
+
+		const queryEntries = Object.entries( args.query );
+
+		// Upon 'GET' args will become part of get params.
+		if ( 'get' === type && queryEntries.length ) {
+			endPoint += '?';
+
+			queryEntries.forEach( ( [ name, value ] ) => {
+				endPoint += name + '=' + value + '&';
+			} );
+		}
+
+		args.query.type = type;
+
+		return endPoint;
+	}
+
+	endpointToCommand( endpoint, query ) {
+		let commandFound = !! $e.data.commands[ endpoint ];
+
+		// Assuming the command maybe index.
+		if ( ! commandFound && $e.data.commands[ endpoint + '/index' ] ) {
+			endpoint = endpoint + '/index';
+			commandFound = true;
+		}
+
+		// Maybe the endpoint includes 'id'. as part of the endpoint.
+		if ( ! commandFound && 'get' === query.type ) {
+			const endpointParts = endpoint.split( '/' ),
+				assumedCommand = endpointParts[ 0 ] + '/' + endpointParts[ 1 ];
+
+			if ( $e.data.commands[ assumedCommand ] ) {
+				endpoint = assumedCommand;
+
+				// Warp with 'id'.
+				query.id = endpointParts[ 2 ];
+
+				commandFound = true;
+			}
+		}
+
+		return endpoint;
+	}
+
+	fetch( type, requestData ) {
+		// TODO: Check if 'include' is required.
+		const params = {
+				credentials: 'include', // cookies
```

Comment:
Add permissions and avoid access if not `Document::is_editable_by_current_user()`